databricks_blueprints/download_file_from_dbfs.py

```diff
@@ -0,0 +1,139 @@
+                if is_binary:
+                    file.write(data)
+                else:
+                    file.write(data.decode('utf-8'))
+                # check if there is still more data left to read
+                if response_data['bytes_read'] == ONE_MB:
+                    offset += ONE_MB
+                else: # that was the last of the data in the file
+                    break
+            elif read_response.status_code == 404:
+                print(f"Error: No file located at {source_file_path}")
+                sys.exit(errors.EXIT_CODE_DBFS_INVALID_FILEPATH)
+            elif read_response.status_code == 400: # Bad Request
+                error = read_response.json()
+                code = error['error_code']
+                message = error['message']
+                print(f"Error: {code} {message}")
+                sys.exit(errors.EXIT_CODE_DBFS_READ_ERROR)
+            else:
+                print(f"Failed to read data: {read_response.status_code} {read_response.text}")
+                sys.exit(errors.EXIT_CODE_DBFS_READ_ERROR)
+    print(f"finished downloading file:{source_file_path} as {dest_file_path}")
+    
+
+def main():
+    args = get_args()
+    access_token = args.access_token
+    instance_url = args.instance_url
+    source_file_name = args.source_file_name
+    source_folder_name = args.source_folder_name
+    dest_file_name = args.dest_file_name
+    dest_folder_name = args.dest_folder_name
+    source_file_name_match_type = args.source_file_name_match_type
+    # create client
+    client = databricks_client.DatabricksClient(access_token, instance_url)
+    
+    if not source_folder_name:
+        source_folder_name = '/FileStore/'
+    else:
+        # get the cleaned file name
+        source_folder_name = '/' + shipyard.files.clean_folder_name(source_folder_name) + '/'
+
+    if not dest_file_name:
+        dest_file_name = source_file_name
+
+    if not dest_folder_name:
+        dest_folder_name = os.getcwd()
+        
+    # get list of all potential file matches
+    if source_file_name_match_type == 'regex_match':
+        files = databricks_client.list_dbfs_files(client, source_folder_name)
+        matching_file_names = shipyard.files.find_all_file_matches(files,
+                                            re.compile(source_file_name))
+        num_matches = len(matching_file_names)
+        if num_matches == 0:
+            print(f"No regex matches found for {source_file_name}")
+            sys.exit(errors.EXIT_CODE_DBFS_INVALID_FILEPATH)
+        print(f'{num_matches} files found. Preparing to download...')
+        for index, file_name in enumerate(matching_file_names):
+            source_file_path = file_name
+            # create destination file name
+            dest_file_name = shipyard.files.extract_file_name_from_source_full_path(file_name)
+            dest_file_name = shipyard.files.enumerate_destination_file_name(dest_file_name, index)
```

Comment:
enumeration should only happen if:
- dest-file-name is provided AND
- more than one file is matched to

See - https://www.shipyardapp.com/docs/reference/blueprints/blueprint-library/match-type/#effects-on-destination-1